core/src/test/java/org/apache/iceberg/rest/RESTCatalogServlet.java

```diff
@@ -0,0 +1,231 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.io.CharStreams;
+import org.apache.iceberg.rest.RESTCatalogAdapter.HTTPMethod;
+import org.apache.iceberg.rest.RESTCatalogAdapter.Route;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The RESTCatalogServlet provides a servlet implementation used in combination with a
+ * RESTCatalogAdaptor to proxy the REST Spec to any Catalog implementation.
+ */
+public class RESTCatalogServlet extends HttpServlet {
+  private static final Logger LOG = LoggerFactory.getLogger(RESTCatalogServlet.class);
+
+  private final RESTCatalogAdapter restCatalogAdapter;
+  private final Map<String, String> responseHeaders =
+      ImmutableMap.of(HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType());
+
+  public RESTCatalogServlet(RESTCatalogAdapter restCatalogAdapter) {
+    this.restCatalogAdapter = restCatalogAdapter;
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    execute(ServletRequestContext.from(request), response);
+  }
+
+  @Override
+  protected void doHead(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    execute(ServletRequestContext.from(request), response);
+  }
+
+  @Override
+  protected void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    execute(ServletRequestContext.from(request), response);
+  }
+
+  @Override
+  protected void doDelete(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    execute(ServletRequestContext.from(request), response);
+  }
+
+  protected void execute(ServletRequestContext context, HttpServletResponse response)
+      throws IOException {
+    response.setStatus(HttpServletResponse.SC_OK);
+    responseHeaders.forEach(response::setHeader);
+
+    if (context.error().isPresent()) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      RESTObjectMapper.mapper().writeValue(response.getWriter(), context.error().get());
+      return;
+    }
+
+    try {
+      Object responseBody =
+          restCatalogAdapter.execute(
+              context.method(),
+              context.path(),
+              context.queryParams(),
+              context.body(),
+              context.route().getResponseClass(),
+              context.headers(),
+              handle(response));
+
+      if (responseBody != null) {
+        RESTObjectMapper.mapper().writeValue(response.getWriter(), responseBody);
+      }
+    } catch (RESTException e) {
+      LOG.error("Error processing REST request", e);
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    } catch (Exception e) {
```

Comment:
Where is this thrown? The only exception I see thrown from `execute` is `RESTException`

Reply:
That is true, but since this is a servlet implementation, this is a safeguard to make sure we return a response (instead of having an empty response) if anything goes wrong during server-side processing.